Data

miretchin opened the pull request:
Added a new `Dataset.batch()` and `.view` function. It uses `functools.partial` so that we can give the batch function a list of keys on which attributes to batch over. It maintains the order of that input for the output.

I also augmented `lookup` to handle `extra` properly, and to handle cases with multiple datapoints per individual compound / SMILES.

Finally, it has a `group` flag, which ensures that the dataloader returned can batch the points by compound / SMILES, rather than a fixed number of observations.

Example usage:

```
from torch.nn.utils.rnn import pad_sequence

def zero_pad(tensor):
    return pad_sequence(
        tensor,
        batch_first=True,
        padding_value=0.0
    ).squeeze()

def get_data(key):
    return list(data.view(group='smiles', by=key))

c, y = get_data('c'), get_data('y')
c_pad, y_pad = zero_pad(c), zero_pad(y)
```

miretchin committed:
Adding pIC50 dataset.

miretchin committed:
New data structure that has a more flexible atch and iew for use in multi-output training loops.

yuanqing-wang commented on malt/data/dataset.py:
```diff
@@ -69,7 +74,7 @@ def __getitem__(self, idx):
             return self.points[idx]
         elif isinstance(idx, str):
-            return self.lookup[idx]
+            return self.__class__(points=self.lookup[idx])
```
sorry don't quite get this bit: if the user feed a SMILES we should return that point, rather than another `Dataset` object?

yuanqing-wang replied:
I don't think the above is in conflict with your `defaultdict` design

miretchin replied:
@yuanqing-wang the fundamental issue is that a dataset can contain multiple datapoints with the same smiles. what then is the appropriate behavior? At least, it should return a list of datapoints. But it seems to me that it may be appropriate to wrap that in a dataset, the same way that if we index several points by `int`s, we also return a new Dataset object.

yuanqing-wang commented on malt/data/dataset.py:
```diff
@@ -159,44 +164,91 @@ def featurize_all(self):
-    def batch_of_g(points):
-        # initialize results
-        gs = []
+    def batch(self, points=None, by=['g', 'y']):
```
I like this design! but have you tested this? not sure if this would work if this is not a `staticmethod` and you need to feed into that torch.dataloader

yuanqing-wang replied:
I think it's safe to make this a `staticmethod` by just removing the dependency on `self` since `self` is never used

miretchin replied:
Tested it. It does work. Sure, I can make it static. I think at some point I was using `self`, but no longer.

miretchin replied:
oh, it does use `self`. if no molecules are provided, it uses self to populate molecules. should I remove that bit?

yuanqing-wang replied:
yeah let's use `@staticmethod` and no self

yuanqing-wang replied:
by definition, `collate_fn` that you feed into a `dataloader` should be static

miretchin replied:
okay. so then we _always_ require that molecules are provided to the `collate_fn`.

miretchin replied:
I made the `batch` function as we know it a private function for use in the dataloader, and made `batch` that is a class function, so that we can still call `data.batch(by=['g', 'y'])`

yuanqing-wang commented:
Also could you add unit tests to these changes?

miretchin committed:
Adding a little interactive playground for understanding use-cases of our new Molecule data structure.

miretchin committed:
Generalized the Point object into a Molecule object.

miretchin committed:
Generalizing the Molecule, adding slicing, enforcing values of same length.

miretchin committed:
Made Molecule base class less opinionated. Just accepts a dictionary and that's all.

miretchin committed:
Further generalize the Molecule base class.

miretchin committed:
Fixed batch to work with new Molecule class.

miretchin committed:
Removed the commented out __getitem__ that was old.

miretchin commented:
I added a new Molecule class. It includes a SMILES, graph `g`, and `metadata`. It assumes that metadata has keys corresponding to different assays, and the corresponding value are records (list of dictionaries).

e.g.

```
molecule['fluorescence']
```
```
[{'c': 7.72e-07, 'y': -0.0486},
 {'c': 3.86e-07, 'y': 0.0359},
 {'c': 1.2e-08, 'y': -0.0295},
 {'c': 9.9e-08, 'y': 0.0575},
 {'c': 0.000198, 'y': -0.256},
 {'c': 9.9e-05, 'y': -0.114}]
```

I also refactored the `batch` function in the Dataset class.

miretchin committed:
Generalized dataset to handle either a Molecule or AssayedMolecule (only difference is the batching function, which comes from malt.data.utils.

yuanqing-wang commented:
as we discussed, it would be great if we do not impose _any_ constraint on the shape of the metadata as a general parent class but provide child classes that caters to our needs

miretchin commented:
> as we discussed, it would be great if we do not impose any constraint on the shape of the metadata as a general parent class but provide child classes that caters to our needs

Done - added a totally unconstrained parent class.

miretchin committed:
Added tests that pass.

miretchin commented:
> Also could you add unit tests to these changes?

Done
